packages/oracle-node/src/fetchers/evm-chain/avalanche/evm-fetcher/sources/curve/CurveRequestHandlers.ts

```
@@ -0,0 +1,112 @@
+      functionsNamesWithValues
+    );
+    return [...erc20BalanceRequests, ...totalSupplyRequest];
+  }
+
+  prepareInternalRequestForCurveTokens = (id: CurveTokensDetailsKeys) => {
+    const { avWBTCAddress, avWETHAddress, av3CRVAddress } =
+      curveTokensContractsDetails[id];
+
+    const wbtcBalanceRequest = this.buildErc20BalanceOfRequest(
+      id,
+      avWBTCAddress
+    );
+    const wethBalanceRequest = this.buildErc20BalanceOfRequest(
+      id,
+      avWETHAddress
+    );
+    const crvBalanceRequest = this.buildErc20BalanceOfRequest(
+      id,
+      av3CRVAddress
+    );
+    return [...wbtcBalanceRequest, ...wethBalanceRequest, ...crvBalanceRequest];
+  };
+
+  buildErc20BalanceOfRequest(
+    id: CurveTokensDetailsKeys,
+    contractAddress: string
+  ) {
+    const { abi, contractWithBalancesAddress } =
+      curveTokensContractsDetails[id];
+
+    return buildMulticallRequests(abi, contractAddress, [
+      { name: "balanceOf", values: [contractWithBalancesAddress] },
+    ]);
+  }
+
+  extractPrice(
+    response: MulticallParsedResponses,
+    id: CurveTokensDetailsKeys
+  ): number | undefined {
+    const { address } = curveTokensContractsDetails[id];
+    const totalSupply = new Decimal(
+      extractValueFromMulticallResponse(response, address, "totalSupply")
+    );
+    const { wbtcLiquidity, wethLiquidity, wcrvLiquidity } =
+      this.calculateLiquiditiesForCurveToken(response, id);
+
+    if (wbtcLiquidity && wethLiquidity && wcrvLiquidity) {
+      const totalLiquidity = wbtcLiquidity
+        .add(wethLiquidity)
+        .add(wcrvLiquidity);
+      return totalLiquidity.div(totalSupply).toNumber();
+    }
+  }
+
+  calculateLiquiditiesForCurveToken(
+    multicallResult: MulticallParsedResponses,
+    id: CurveTokensDetailsKeys
+  ) {
+    const { avWBTCAddress, avWETHAddress, av3CRVAddress, tokensToFetch } =
+      curveTokensContractsDetails[id];
+    const tokenToFetchPrice = getTokensPricesFromLocalCache(tokensToFetch);
```

If we fail fast here we don't have to check #here1?

---

Ok, I added check for last price in `getTokensPricesFromLocalCache`